File: MiniGame-Java/src/main/java/rs/lab/mges/engine/Sound.java
package rs.lab.mges.engine;

import com.sun.jna.Pointer;
import com.sun.jna.ptr.IntByReference;
import com.sun.jna.ptr.PointerByReference;
import io.github.libsdl4j.api.audio.SDL_AudioDeviceID;
import io.github.libsdl4j.api.audio.SDL_AudioFormat;
import io.github.libsdl4j.api.audio.SDL_AudioSpec;
import io.github.libsdl4j.api.audio.SdlAudio;
import java.util.ArrayList;
import java.util.Collection;

import static io.github.libsdl4j.api.Sdl.SDL_WasInit;
import static io.github.libsdl4j.api.SdlSubSystemConst.SDL_INIT_AUDIO;
import static io.github.libsdl4j.api.audio.SdlAudioConst.AUDIO_S16LSB;
import static io.github.libsdl4j.api.audio.SdlAudioConst.SDL_MIX_MAXVOLUME;
import static io.github.libsdl4j.api.error.SdlError.SDL_GetError;
import static io.github.libsdl4j.api.audio.SdlAudio.SDL_CloseAudioDevice;
import static io.github.libsdl4j.api.audio.SdlAudio.SDL_FreeWAV;
import static io.github.libsdl4j.api.audio.SdlAudio.SDL_LockAudioDevice;
import static io.github.libsdl4j.api.audio.SdlAudio.SDL_PauseAudioDevice;
import static io.github.libsdl4j.api.audio.SdlAudio.SDL_UnlockAudioDevice;
import static io.github.libsdl4j.api.audio.SdlAudioConst.SDL_AUDIO_ALLOW_FORMAT_CHANGE;

public class Sound {

    private final static int AUDIO_FREQUENCY = 44100;
    private final static int AUDIO_FORMAT = AUDIO_S16LSB;
    private final static int AUDIO_CHANNELS = 2;
    private final static int AUDIO_SAMPLES = 4096;
    private final static int AUDIO_MAX_SOUNDS = 25;

    private static class PrivateAudioDevice {

        public SDL_AudioDeviceID device;
        public SDL_AudioSpec want;
        public boolean audioEnabled;
    }

    public static class Audio {

        public int length;
        public int lengthTrue;
        public long bufferTrue;
        public long buffer;
        public int position;
        public boolean loop;
        public boolean fade;
        public boolean free;
        public int volume;
        public SDL_AudioSpec audio;
        public boolean active;

        // <editor-fold defaultstate="collapsed" desc="Builder">
        public static Audio builder() {
            return new Audio();
        }

        public Audio setLength(int length) {
            this.length = length;
            return this;
        }

        public Audio setLengthTrue(int lengthTrue) {
            this.lengthTrue = lengthTrue;
            return this;
        }

        public Audio setBufferTrue(long bufferTrue) {
            this.bufferTrue = bufferTrue;
            return this;
        }

        public Audio setBuffer(long buffer) {
            this.buffer = buffer;
            return this;
        }

        public Audio setPosition(int position) {
            this.position = position;
            return this;
        }

        public Audio setLoop(boolean loop) {
            this.loop = loop;
            return this;
        }

        public Audio setFade(boolean fade) {
            this.fade = fade;
            return this;
        }

        public Audio setFree(boolean free) {
            this.free = free;
            return this;
        }

        public Audio setVolume(int volume) {
            this.volume = volume;
            return this;
        }

        public Audio setAudio(SDL_AudioSpec audio) {
            this.audio = audio;
            return this;
        }

        public Audio setActive(boolean active) {
            this.active = active;
            return this;
        }
        // </editor-fold>
    }

    private static final Collection<Audio> audios = new ArrayList<>();
    private static final PrivateAudioDevice device = new PrivateAudioDevice();

    private static int soundCount = 0;

    public static void playSound(String filename, int volume) {
        playAudio(filename, null, false, volume);
    }

    public static void playMusic(String filename) {
        playMusic(filename, SDL_MIX_MAXVOLUME);
    }

    public static void playMusic(String filename, int volume) {
        playAudio(filename, null, true, volume);
    }

    public static void playSoundFromMemory(Audio audio) {
        playSoundFromMemory(audio, SDL_MIX_MAXVOLUME);
    }

    public static void playSoundFromMemory(Audio audio, int volume) {
        playAudio(null, audio, false, volume);
    }

    public static void playMusicFromMemory(Audio audio) {
        playMusicFromMemory(audio, SDL_MIX_MAXVOLUME);
    }

    public static void playMusicFromMemory(Audio audio, int volume) {
        playAudio(null, audio, true, volume);
    }

    public static void initAudio() {
        if ((SDL_WasInit(SDL_INIT_AUDIO) & SDL_INIT_AUDIO) == 0) {
            SDLUtils.LogErr("Error: SDL_INIT_AUDIO not initialized");
            return;
        }
        
        device.want = new SDL_AudioSpec();
        device.want.freq = AUDIO_FREQUENCY;
        device.want.format = new SDL_AudioFormat(AUDIO_FORMAT);
        device.want.channels = AUDIO_CHANNELS;
        device.want.samples = AUDIO_SAMPLES;
        device.want.callback = (Pointer userdata, Pointer stream, int len) -> audioCallback(userdata, stream, len);
        device.want.userdata = null;

        SDL_AudioSpec obtained = new SDL_AudioSpec();
        device.device = SdlAudio.SDL_OpenAudioDevice(null, 0, device.want, obtained, SDL_AUDIO_ALLOW_FORMAT_CHANGE);
        if (device.device.intValue() == 0) {
            SDLUtils.LogInfo("Warning: failed to open audio device: %s", SDL_GetError());
        } else if (device.want.format.longValue() != obtained.format.longValue()) {
            SDLUtils.LogInfo("Warning: format missmatch wanted %d but got %d", device.want.format, obtained.format);
        } else {
            /* Set audio device enabled global flag */
            device.audioEnabled = true;

            /* Unpause active audio stream */
            unpauseAudio();
        }
    }

    public static void audioCallback(Pointer userdata, Pointer stream, int len) {
        /* Silence the main buffer (important!) */
        stream.clear(len);

        /* Mixing audio */
        for (var audio : audios) {
            if (audio.length > 0) {
                var tempLength = Math.min(len, audio.length);
                var ptr = Pointer.createConstant(audio.buffer + audio.position);
                SdlAudio.SDL_MixAudioFormat(stream, ptr, audio.audio.format, tempLength, audio.volume);
                audio.position += tempLength;
                audio.length -= tempLength;
            } else {
                if (audio.loop) {
                    audio.buffer = audio.bufferTrue;
                    audio.length = audio.lengthTrue;
                    audio.position = 0;
                } else {
                    if (soundCount > 0) {
                        soundCount--;
                    }
                }
            }
        }
    }

    public static void endAudio() {
        if (device.audioEnabled) {
            pauseAudio();

            SDL_LockAudioDevice(device.device);
            /* Close down audio */
            SDL_CloseAudioDevice(device.device);
        }
    }

    public static void pauseAudio() {
        if (device.audioEnabled) {
            SDL_PauseAudioDevice(device.device, 1);
        }
    }

    public static void unpauseAudio() {
        if (device.audioEnabled) {
            SDL_PauseAudioDevice(device.device, 0);
        }
    }

    public static void freeAudio(Audio audio) {
        if (audio.free) {
            SDL_FreeWAV(Pointer.createConstant(audio.bufferTrue));
        }
    }

    public static Audio createAudio(String filename, boolean loop, int volume) {
        Audio new_ = new Audio();

        if (Helpers.isNullOrEmpty(filename)) {
            SDLUtils.LogInfo("Warning: filename empty");
            return null;
        }

        new_.loop = loop;
        new_.fade = false;
        new_.free = true;
        new_.volume = volume;
        new_.audio = new SDL_AudioSpec();

        PointerByReference buffer = new PointerByReference();
        IntByReference bufferLen = new IntByReference();

        if (SdlAudio.SDL_LoadWAV(filename, new_.audio, buffer, bufferLen) == null) {
            SDLUtils.LogInfo("Warning: failed to open wave file: %s error: %s", filename, SDL_GetError());
            return null;
        }

        // Pointer-to-pointer: return the real buffer pointer
        new_.bufferTrue = Pointer.nativeValue(buffer.getPointer().getPointer(0));
        new_.lengthTrue = bufferLen.getValue();

        if (device.want.format.longValue() != new_.audio.format.longValue()) {
            SDLUtils.LogInfo("Warning: format missmatch wanted 0x%04x but got 0x%04x", device.want.format, new_.audio.format);
        }
        new_.buffer = new_.bufferTrue;
        new_.length = new_.lengthTrue;
        new_.position = 0;
        new_.active = true;

        return new_;
    }

    public static void playSound(String filename) {
        playAudio(filename, null, false, SDL_MIX_MAXVOLUME);
    }

    private static void playAudio(String filename, Audio audio, boolean loop) {
        playAudio(filename, audio, loop, SDL_MIX_MAXVOLUME);
    }

    private static void playAudio(String filename, Audio audio, boolean loop, int volume) {
        Audio new_;
        /* Check if audio is enabled */
        if (!device.audioEnabled) {
            return;
        }

        /* If sound, check if under max number of sounds allowed, else don't play */
        if (loop == false) {
            if (soundCount >= AUDIO_MAX_SOUNDS) {
                return;
            } else {
                soundCount++;
            }
        }
        /* Load from filename or from Memory */
        if (!Helpers.isNullOrEmpty(filename)) {
            /* Create new music sound with loop */
            var nnew_ = createAudio(filename, loop, volume);
            if (nnew_ == null) {
                return;
            }

            new_ = nnew_;
        } else if (audio != null) {
            new_ = Audio.builder()
                    .setLoop(loop)
                    .setFade(false)
                    .setVolume(volume)
                    .setPosition(0)
                    .setBuffer(audio.buffer)
                    .setLength(audio.length)
                    .setBufferTrue(audio.buffer)
                    .setLengthTrue(audio.length)
                    .setActive(true)
                    .setFree(false)
                    .setAudio(audio.audio);
        } else {
            SDLUtils.LogInfo("Warning: filename and Audio parameters NULL");
            return;
        }

        /* Lock callback function */
        SdlAudio.SDL_LockAudioDevice(device.device);

        audios.add(new_);

        SDL_UnlockAudioDevice(device.device);
    }

}

File: MiniGame-NET/MiniGameEngine/Sound.cs
// Copyright 2023 Peter Bakota
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using MiniGameEngine.Utils;
using SDL2;

namespace MiniGameEngine
{
    public static class Sound
    {
        private const int AUDIO_FREQUENCY = 44100;
        private const ushort AUDIO_FORMAT = SDL.AUDIO_S16LSB;
        private const byte AUDIO_CHANNELS = 2;
        private const ushort AUDIO_SAMPLES = 4096;
        private const int AUDIO_MAX_SOUNDS = 25;

        private sealed record PrivateAudioDevice
        {
            public uint device;
            public SDL.SDL_AudioSpec want;
            public bool audioEnabled;
        }

        public record Audio
        {
            public uint length;
            public uint lengthTrue;
            public IntPtr bufferTrue;
            public IntPtr buffer;
            public bool loop;
            public bool fade;
            public bool free;
            public int volume;
            public SDL.SDL_AudioSpec audio;
            public bool Active { get; set; }
        };

        private static readonly List<Audio> _audios = new();
        private static readonly PrivateAudioDevice _device = new();
        private static uint _soundCount = 0;

        public static void PlaySound(string filename, int volume = SDL.SDL_MIX_MAXVOLUME)
            => PlayAudio(filename, null, false, volume);

        public static void PlayMusic(string filename, int volume = SDL.SDL_MIX_MAXVOLUME)
            => PlayAudio(filename, null, true, volume);

        public static void PlaySoundFromMemory(Audio audio, int volume = SDL.SDL_MIX_MAXVOLUME)
            => PlayAudio(null, audio, false, volume);

        public static void PlayMusicFromMemory(Audio audio, int volume = SDL.SDL_MIX_MAXVOLUME)
            => PlayAudio(null, audio, true, volume);

        public static void InitAudio()
        {
            if ((SDL.SDL_WasInit(SDL.SDL_INIT_AUDIO) & SDL.SDL_INIT_AUDIO) == 0)
            {
                SDLUtil.LogErr("Error: SDL_INIT_AUDIO not initialized");
                return;
            }

            _device.want.freq = AUDIO_FREQUENCY;
            _device.want.format = AUDIO_FORMAT;
            _device.want.channels = AUDIO_CHANNELS;
            _device.want.samples = AUDIO_SAMPLES;
            _device.want.callback = SoundCallback;
            _device.want.userdata = IntPtr.Zero;

            _device.device = SDL.SDL_OpenAudioDevice(IntPtr.Zero, 0, ref _device.want, out var obtained, (int)SDL.SDL_AUDIO_ALLOW_FORMAT_CHANGE);
            if (_device.device == 0)
            {
                SDLUtil.LogInfo($"Warning: failed to open audio device: {SDL.SDL_GetError()}");
            }
            else if (_device.want.format != obtained.format)
            {
                SDLUtil.LogInfo($"Warning: format missmatch wanted {_device.want.format} but got {obtained.format}");
            }
            else
            {
                /* Set audio device enabled global flag */
                _device.audioEnabled = true;

                /* Unpause active audio stream */
                UnpauseAudio();
            }
        }

        private static void SoundCallback(IntPtr userdata, IntPtr stream, int len)
        {
            /* Silence the main buffer (important!) */
            SDL.SDL_memset(stream, 0, (IntPtr)len);

            foreach (var audio in _audios)
            {
                // TODO: fade in/out support
                if (audio.length > 0)
                {
                    var tempLength = len > audio.length ? (int)audio.length : len;
                    SDL.SDL_MixAudioFormat(stream, audio.buffer, AUDIO_FORMAT, (uint)tempLength, audio.volume);
                    audio.buffer += tempLength;
                    audio.length -= (uint)tempLength;
                }
                else
                {
                    if (audio.loop)
                    {
                        audio.buffer = audio.bufferTrue;
                        audio.length = audio.lengthTrue;
                    }
                    else
                    {
                        audio.Active = false;
                        if (_soundCount > 0)
                            _soundCount--;
                    }
                }
            }
        }

        public static void EndAudio()
        {
            if (_device.audioEnabled)
            {
                PauseAudio();

                /* Close down audio */
                SDL.SDL_CloseAudioDevice(_device.device);
            }
        }

        public static void PauseAudio()
        {
            if (_device.audioEnabled)
                SDL.SDL_PauseAudioDevice(_device.device, 1);
        }

        public static void UnpauseAudio()
        {
            if (_device.audioEnabled)
                SDL.SDL_PauseAudioDevice(_device.device, 0);
        }


        public static void FreeAudio(Audio audio)
        {
            SDL.SDL_FreeWAV(audio.bufferTrue);
        }

        private static Audio? CreateAudio(string? filename, bool loop, int volume)
        {
            Audio new_ = new();

            if (string.IsNullOrEmpty(filename))
            {
                SDLUtil.LogInfo($"Warning: filename NULL: {filename}");
                return null;
            }

            new_.loop = loop;
            new_.fade = false;
            new_.free = true;
            new_.volume = volume;

            if (SDL.SDL_LoadWAV(filename, out new_.audio, out new_.bufferTrue, out new_.lengthTrue) == IntPtr.Zero)
            {
                SDLUtil.LogInfo($"Warning: failed to open wave file: {filename} error: {SDL.SDL_GetError()}");
                return null;
            }

            if (_device.want.format != new_.audio.format)
            {
                SDLUtil.LogInfo($"Warning: format missmatch wanted {_device.want.format} but got {new_.audio.format}");
            }

            new_.buffer = new_.bufferTrue;
            new_.length = new_.lengthTrue;
            new_.Active = true;

            return new_;
        }

        private static void PlayAudio(string? filename, Audio? audio, bool loop, int volume = SDL.SDL_MIX_MAXVOLUME)
        {
            Audio new_;

            /* Check if audio is enabled */
            if (!_device.audioEnabled)
                return;

            /* If sound, check if under max number of sounds allowed, else don't play */
            if (!loop && _soundCount >= AUDIO_MAX_SOUNDS)
                return;

            /* Load from filename or from Memory */
            if (!string.IsNullOrEmpty(filename))
            {
                /* Create new music sound with loop */
                var nnew_ = CreateAudio(filename, loop, volume);
                if (nnew_ == null) return;

                new_ = nnew_;
            }
            else if (audio != null)
            {
                new_ = new()
                {
                    loop = loop,
                    fade = false,
                    volume = volume,
                    buffer = audio.buffer,
                    length = audio.length,
                    bufferTrue = audio.buffer,
                    lengthTrue = audio.length,
                    Active = true,
                    free = false
                };
            }
            else
            {
                SDLUtil.LogInfo("Warning: filename and Audio parameters NULL");
                return;
            }

            /* Lock callback function */
            SDL.SDL_LockAudioDevice(_device.device);

            _audios.RemoveAll(s => !s.Active);
            _audios.Add(new_);

            SDL.SDL_UnlockAudioDevice(_device.device);

            if(!loop) {
                _soundCount++;
            }
        }
    }
}